model/task/generate_tasks_estimations.go

```diff
@@ -0,0 +1,118 @@
+package task
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/utility"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const (
+	lookBackTime = 7 * 24 * time.Hour // one week
+)
+
+type generateTasksEstimationsResults struct {
+	DisplayName        string  `bson:"_id"`
+	EstimatedCreated   float64 `bson:"est_created"`
+	EstimatedActivated float64 `bson:"est_activated"`
+}
+
+func (t *Task) setGenerateTasksEstimations() error {
+	// Do not run if estimations have already been cached.
+	if t.EstimatedNumGeneratedTasks != nil && t.EstimatedNumActivatedGeneratedTasks != nil {
+		return nil
+	}
+
+	match := bson.M{
+		BuildVariantKey:   t.BuildVariant,
+		ProjectKey:        t.Project,
+		DisplayNameKey:    t.DisplayName,
+		GeneratedTasksKey: true,
+		StatusKey: bson.M{
+			"$in": evergreen.TaskCompletedStatuses,
+		},
+		StartTimeKey: bson.M{
+			"$gt": time.Now().Add(-1 * lookBackTime),
+		},
+		FinishTimeKey: bson.M{
+			"$lte": time.Now(),
+		},
+	}
+
+	pipeline := []bson.M{
+		{
+			"$match": match,
+		},
+		{
+			"$project": bson.M{
+				DisplayNameKey:                1,
+				NumGeneratedTasksKey:          1,
+				NumActivatedGeneratedTasksKey: 1,
+				IdKey:                         0,
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id": fmt.Sprintf("$%s", DisplayNameKey),
+				"est_created": bson.M{
+					"$avg": fmt.Sprintf("$%s", NumGeneratedTasksKey),
+				},
+				"est_activated": bson.M{
+					"$avg": fmt.Sprintf("$%s", NumActivatedGeneratedTasksKey),
+				},
+			},
+		},
+	}
+
+	// anonymous struct for unmarshalling result bson
+	results := []generateTasksEstimationsResults{}
+
+	coll := evergreen.GetEnvironment().DB().Collection(Collection)
+	ctx, cancel := evergreen.GetEnvironment().Context()
```

Comment:
Avoid using the env context here if possible because it's basically the same as `context.Background()` - the aggregation can run forever until the app server shuts down. That means if the query is expensive and is taking a long time, nothing will ever stop it from just running forever.

Typically, you should use the context passed into the function (which should be passed down from the source, like an Amboy job's `Run` context), so requiring a context be passed into `setGenerateTasksEstimations` would make things safer.